Using enum instead of boolean to filter analyses in list-analysis

haverma opened the pull request:
- Using boolean caused the client to always get only the user policies even when it passed `null` for `suggested` parameter, as  `null` was converted to `false` when received by `WorkMgrService`

- Enum will prevent `nulls` being exchanged between the client and service when providing a valid value of `AnalysisType` is made mandatory for client in the future

dhalperi commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1212 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVXQaNBa0hLI6BkZK:-L6SVXQaNBa0hLI6BkZL:b-vxy6w1) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1212)):*
> ```Java
>                 .filter(
>                     aName -> {
>                       if (analysisType == AnalysisType.ALL || analysisType == null) {
> ```

simplify this by earlier doing `analysisType = firstNonNull(analysisType, AnalysisType.ALL)`

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1216 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVjGip9jRUOPGQd0N:-L6SVjGip9jRUOPGQd0O:b-9mwgui) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1216)):*
> ```Java
>                       }
>                       if (analysisType == AnalysisType.SUGGESTED
>                           && AnalysisMetadataMgr.getAnalysisSuggestedOrFalse(
> ```

cache this result so we don't have to make multiple calls to the metadata manager.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1225 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVpqLzs4l0nSZ2WEB:-L6SVpqLzs4l0nSZ2WEC:bknth8s) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1225)):*
> ```Java
>                         return true;
>                       }
>                       return false;
> ```

pull this out in a separate private helper function where the logic is clearer.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java, line 1063 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVz6XdRf2jObzD3RO:-L6SVz6XdRf2jObzD3RP:b-qhlybz) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java#L1063)):*
> ```Java
>       @FormDataParam(CoordConsts.SVC_KEY_VERSION) String clientVersion,
>       @FormDataParam(CoordConsts.SVC_KEY_CONTAINER_NAME) String containerName,
>       @Nullable @FormDataParam(CoordConsts.SVC_KEY_ANALYSIS_TYPE) AnalysisType analysisType) {
> ```

what clients, if any, need to be updated? Are those PRs in-flight?

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkMgrTest.java, line 219 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SWAA4iUXGHyp75Sj4:-L6SWAA5ux903Ro4W92n:b-wka5f) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/test/java/org/batfish/coordinator/WorkMgrTest.java#L219)):*
> ```Java
>     analyses = _manager.listAnalyses(containerName, AnalysisType.ALL);
> 
>     // checking that both analyses are returned for AnalysisType as ALL
> ```

nice. But, in this test the whitespace just confuses thing. I would honestly drop the local `analyses` variables:

```
assertThat(_manager.listAnalyses(..., ALL), equalTo(Sets.newHashSet("a1", "a2")));
assertThat(_manager.listAnalyses(..., SUGGESTED), equalTo(Sets.newHashSet("a1")));
assertThat(_manager.listAnalyses(..., USER), equalTo(Sets.newHashSet("a2")));
```

No need to use immutablesortedset in tests, all that matters for the test is that it be some kind of set. Use the clearest and most concise `Set` constructor you can, which I suggest may be the above but there may be better ones.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

corinaminer commented:
Reviewed 2 of 5 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1212 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVXQaNBa0hLI6BkZK:-L6SWoSZQWOpNvHRwbOv:b-flp0tn) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1212)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

simplify this by earlier doing `analysisType = firstNonNull(analysisType, AnalysisType.ALL)`
</blockquote></details>

actually i think that we'll want a null input for analysisType to return only user policies, since that's what the other clients would expect to get.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1212 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVXQaNBa0hLI6BkZK:-L6SodzNQ9YlJvMOh7ZU:b-ri3ta3) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1212)):*
<details><summary><i>Previously, corinaminer (Corina Miner) wrote…</i></summary><blockquote>

actually i think that we'll want a null input for analysisType to return only user policies, since that's what the other clients would expect to get.
</blockquote></details>

Makes sense @corinaminer 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 2 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1212 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVXQaNBa0hLI6BkZK:-L6UvsVjo_mvQG06P-i-:bjmtpij) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1212)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Makes sense @corinaminer 
</blockquote></details>

done in `WorkMgrService#listAnalyses`

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1216 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVjGip9jRUOPGQd0N:-L6Uw2R4DJRliwdQqlnR:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1216)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

cache this result so we don't have to make multiple calls to the metadata manager.
</blockquote></details>

done

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1225 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVpqLzs4l0nSZ2WEB:-L6Uw30D6LnkGTsyp_pc:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1225)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

pull this out in a separate private helper function where the logic is clearer.
</blockquote></details>

done

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java, line 1063 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVz6XdRf2jObzD3RO:-L6Uw4WRjzhaaIIMMGXg:b-gurmfu) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java#L1063)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

what clients, if any, need to be updated? Are those PRs in-flight?
</blockquote></details>

`Null`is a valid input for analysisType, so no clients will need to be updated. But I will update the ui client as a part of the PR submitted for create default policies

---

*[projects/coordinator/src/test/java/org/batfish/coordinator/WorkMgrTest.java, line 219 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SWAA4iUXGHyp75Sj4:-L6UwXaLdOjczMUIDcX9:bb7f2mf) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/test/java/org/batfish/coordinator/WorkMgrTest.java#L219)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

nice. But, in this test the whitespace just confuses thing. I would honestly drop the local `analyses` variables:

```
assertThat(_manager.listAnalyses(..., ALL), equalTo(Sets.newHashSet("a1", "a2")));
assertThat(_manager.listAnalyses(..., SUGGESTED), equalTo(Sets.newHashSet("a1")));
assertThat(_manager.listAnalyses(..., USER), equalTo(Sets.newHashSet("a2")));
```

No need to use immutablesortedset in tests, all that matters for the test is that it be some kind of set. Use the clearest and most concise `Set` constructor you can, which I suggest may be the above but there may be better ones.
</blockquote></details>

done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1216 at r1](https://beta.reviewable.io/reviews/batfish/batfish/998#-L6SVjGip9jRUOPGQd0N:-L6XcIwY388xwCSeuMbl:bxpwxpf) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1216)):*
<details><summary><i>Previously, haverma (Harsh Verma) wrote…</i></summary><blockquote>

done
</blockquote></details>

What I said cache I just literally meant in a local variable in this function. Is there a reason we need a long-lived cache?

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 104 at r2](https://beta.reviewable.io/reviews/batfish/batfish/998#-L6XbzO11PjFIgY8BnHY:-L6XbzO11PjFIgY8BnHZ:bb7phvt) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L104)):*
> ```Java
>   }
> 
>   private Cache<String, Boolean> _cacheSuggedted;
> ```

typo

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1210 at r2](https://beta.reviewable.io/reviews/batfish/batfish/998#-L6XcBbT7wGnSvrEg3v7:-L6XcBbT7wGnSvrEg3v8:b-caljsz) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1210)):*
> ```Java
>     Path analysesDir = containerDir.resolve(BfConsts.RELPATH_ANALYSES_DIR);
>     if (!Files.exists(analysesDir)) {
>       return new TreeSet<>();
> ```

Does the return value here need to be mutable? If not, use ImmutableSortedSet here and below.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1217 at r2](https://beta.reviewable.io/reviews/batfish/batfish/998#-L6Xc4GS1uvAJV2YtvhM:-L6Xc4GS1uvAJV2YtvhN:bss0m9c) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1217)):*
> ```Java
>             .map(subdir -> subdir.getFileName().toString())
>             .filter(aName -> selectAnalysis(aName, analysisType, containerName))
>             .distinct()
> ```

what is the distinct for? How can we get duplicate filenames?

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java, line 1063 at r1](https://beta.reviewable.io/reviews/batfish/batfish/998#-L6SVz6XdRf2jObzD3RO:-L6XhVSLCDfEfHP7KoNZ:bidbbjk) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java#L1063)):*
<details><summary><i>Previously, haverma (Harsh Verma) wrote…</i></summary><blockquote>

`Null`is a valid input for analysisType, so no clients will need to be updated. But I will update the ui client as a part of the PR submitted for create default policies
</blockquote></details>

What about clients that used to send `true` or `false`?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

haverma commented:
Review status: 4 of 5 files reviewed at latest revision, 5 unresolved discussions.

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1216 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVjGip9jRUOPGQd0N:-L6Y5eHe9vTsM81n6OLd:b-rcfkz0) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1216)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

What I said cache I just literally meant in a local variable in this function. Is there a reason we need a long-lived cache?
</blockquote></details>

Using a local variable now,  I misunderstood that you meant reducing actual calls to metadata manager between different listAnalyses calls. 

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 104 at r2](https://reviewable.io:443/reviews/batfish/batfish/998#-L6XbzO11PjFIgY8BnHY:-L6Y5KfiHJ0YxbYbXnUY:b-4gijag) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L104)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

typo
</blockquote></details>

removed cache, explaining in other comment

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1210 at r2](https://reviewable.io:443/reviews/batfish/batfish/998#-L6XcBbT7wGnSvrEg3v7:-L6Y5QTa3tcCZ84MPeq9:bjwkd2a) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1210)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

Does the return value here need to be mutable? If not, use ImmutableSortedSet here and below.
</blockquote></details>

No, changed to ImmutableSortedSet

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java, line 1217 at r2](https://reviewable.io:443/reviews/batfish/batfish/998#-L6Xc4GS1uvAJV2YtvhM:-L6Y5RlkqJ5Wk__ed2m2:bvtc0qn) ([raw file](https://github.com/batfish/batfish/blob/cf936cb91c28e336243ee1cbe1ec96a3e711ebd7/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L1217)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

what is the distinct for? How can we get duplicate filenames?
</blockquote></details>

thanks for catching that, it got introduced by intelliJ stream code optimization and I missed it, removed

---

*[projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java, line 1063 at r1](https://reviewable.io:443/reviews/batfish/batfish/998#-L6SVz6XdRf2jObzD3RO:-L6Y65FYx4t9iqxp0WIZ:b17lrji) ([raw file](https://github.com/batfish/batfish/blob/b52c66cffbd85a8e3107e4b40ba03a018ae0d374/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgrService.java#L1063)):*
<details><summary><i>Previously, dhalperi (Dan Halperin) wrote…</i></summary><blockquote>

What about clients that used to send `true` or `false`?
</blockquote></details>

All clients are currently passing Null for this param, they are not including this param at all in the POST request body

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998)*
<!-- Sent from Reviewable.io -->

dhalperi commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/batfish/batfish/998#-:-L6Y7izZAEikngqhEUX_:bnfp4nl)*
<!-- Sent from Reviewable.io -->
